### PR TITLE
ENT-9685: Added selinux policy source files and instructions on how to use them

### PR DESCRIPTION
**Pull request by craigcomstock**
https://github.com/cfengine/core/pull/5146
https://github.com/cfengine/buildscripts/pull/1187

### CODE REVIEW

**Comment by cf-bottom** [comment]
Thank you for submitting a pull request! Maybe @Lex-2008 can review this?

**Comment by craigcomstock** [comment]
try [Build Status](https://ci.cfengine.com/job/pr-pipeline/9134/)

**Comment by craigcomstock** [comment]
retry [Build Status](https://ci.cfengine.com/job/pr-pipeline/9137/)

**Comment by craigcomstock** [comment]
cf-bottom: I wish you could help me out with these, oh well, I'll teach you about it another time. Here's my retry of rhel-9 hub only: [Build Status](https://ci.cfengine.com/job/pr-pipeline/9138/)

**Comment by craigcomstock** [comment]
[Build Status](https://ci.cfengine.com/job/pr-pipeline/9140/)

**Comment by craigcomstock** [comment]
ok! latest build package fails as expected :) 

```
[vagrant@trhel-9 ~]$ sudo rpm -i cfengine-nova-hub-3.22.0a.bc20189be-24835.el9.x86_64.rpm 
error: Failed dependencies:
        libselinux >= 3.4 is needed by cfengine-nova-hub-3.22.0a.bc20189be-24835.el9.x86_64
```

On a fresh unregistered and unattached rhel-9 vagrant libvirt vm: 

```
$ yum list installed | grep libselinux
libselinux.x86_64                             3.3-2.el9                     @anaconda             
libselinux-utils.x86_64                       3.3-2.el9                     @anaconda             
python3-libselinux.x86_64                     3.3-2.el9                     @AppStream    
```

After registering, attaching and upgrading:

```
subscription-manager register
subscription-manager attach
yum upgrade -y
```

I am able to install the package without issue.

File: packaging/cfengine-nova-hub/cfengine-nova-hub.spec.in [hunk 1]
```diff
@@ -12,11 +12,21 @@ License: COSL
...
 
+# we require libselinux version that matches or exceeds our build systems
+# since we compile /var/cfengine/selinux/cfengine-enterprise.pp and that
+# has a particular policydb version which must be accepted by the install host.
+%if %{?rhel}%{!?rhel:0} == 9
+Requires: libselinux >= 3.4
+%endif
+%if %{?rhel}%{!?rhel:0} == 8
+Requires: libselinux >= 2.9
+%endif
```
Comment (vpodzime) [inline packaging/cfengine-nova-hub/cfengine-nova-hub.spec.in:28]: We should detect the version of the required package in the build system, not just have it hardcoded. And I'm not sure `libselinux` is the right package to require:

> libselinux provides an API for SELinux applications to get and set
process and file security contexts and to obtain security policy
decisions.  Required for any applications that use the SELinux API.

What about `selinux-policy-base`? 
```
$ rpm -q --provides selinux-policy
config(selinux-policy) = 37.17-1.fc37
rpm_macro(_file_context_file)
rpm_macro(_file_context_file_pre)
rpm_macro(_file_custom_defined_booleans)
rpm_macro(_file_custom_defined_booleans_tmp)
rpm_macro(_selinux_policy_version)
rpm_macro(_selinux_store_path)
rpm_macro(_selinux_store_policy_path)
rpm_macro(selinux_modules_install)
rpm_macro(selinux_modules_uninstall)
rpm_macro(selinux_relabel_post)
rpm_macro(selinux_relabel_pre)
rpm_macro(selinux_requires)
rpm_macro(selinux_set_booleans)
rpm_macro(selinux_unset_booleans)
selinux-policy = 37.17-1.fc37
selinux-policy-base = 37.17-1.fc37
```

File: packaging/common/cfengine-hub/postinstall.sh [hunk 2]
```diff
@@ -11,7 +11,15 @@ then
...
-    cf_console echo "warning! semodule failed"
+    cf_console echo "warning! semodule import failed, examine /var/log/CFE*log and \
+consider installing selinux-policy-devel package and \
+rebuilding policy with: \
+\
+cd $PREFIX/selinux; make -f /usr/share/selinux/devel/Makefile -j1 \
+\
+and then restarting services with  \
+\
+systemctl restart cfengine3"
```
Comment (vpodzime) [inline packaging/common/cfengine-hub/postinstall.sh:22]: They also need to `semodule -i` the built module.

File: packaging/cfengine-nova-hub/cfengine-nova-hub.spec.in (hunk 1 above)
Comment (craigcomstock) [reply packaging/cfengine-nova-hub/cfengine-nova-hub.spec.in:28]: Agreed. Better package that I imagine is involved if you upgrade libselinux.

We could probably autotools this business. I'll check it out.

File: packaging/common/cfengine-hub/postinstall.sh (hunk 2 above)
Comment (craigcomstock) [reply packaging/common/cfengine-hub/postinstall.sh:22]: Correct. Will add.

**Comment by craigcomstock** [comment]
retry after some changes: [Build Status](https://ci.cfengine.com/job/pr-pipeline/9143/)

File: build-scripts/package [hunk 3]
```diff
@@ -129,9 +129,22 @@ case "$PACKAGING" in
         fi
     fi
 
+    # determine policy-version for rhel systems so we can require what we build with
+    if [ "$OS" = "rhel" ]; then
+        enforcing=$(cat /sys/fs/selinux/enforce)
+        if [ "$enforcing" = "1" ]; then
+            SELINUX_POLICY_VERSION=$(rpm -q selinux-policy | sed -n 's/[a-z-]*\([0-9]*.[0-9]*.[0-9]*\).*$/\1/p')
```
Comment (vpodzime) [inline build-scripts/package]: The system doesn't have to be in the _Enforcing_ mode for this.

Comment (vpodzime) [reply build-scripts/package]:
```
$ rpm -q --qf '%{VERSION}\n' selinux-policy
37.17
```

**Comment by craigcomstock** [comment]
retry after a bit more work and squashing commits: [Build Status](https://ci.cfengine.com/job/pr-pipeline/9154/)

File: packaging/cfengine-nova-hub/cfengine-nova-hub.spec.in [hunk 4]
```diff
@@ -12,11 +12,17 @@ License: COSL
 Group: Applications/System
 URL: http://cfengine.com/
 BuildRoot: %{_topdir}/%{name}-%{version}-%{release}-buildroot
-Obsoletes: cfengine3, cfengine-community
+Obsoletes: cfengine3 < @@VERSION@@, cf-community < @@VERSION@@
```
Comment (vpodzime) [inline packaging/cfengine-nova-hub/cfengine-nova-hub.spec.in:15]: This is an unrelated change that deserves a separate commit with an explanation.

**Review by vpodzime — COMMENTED** [review]
Looks good to me otherwise.

File: packaging/cfengine-nova-hub/cfengine-nova-hub.spec.in (hunk 4 above)
Comment (craigcomstock) [reply packaging/cfengine-nova-hub/cfengine-nova-hub.spec.in:15]: fair. :)

Comment (craigcomstock) [reply packaging/cfengine-nova-hub/cfengine-nova-hub.spec.in:15]: It was related enough because without it I couldn't get the rpm to build. FWIW. :)

Comment (craigcomstock) [reply packaging/cfengine-nova-hub/cfengine-nova-hub.spec.in:15]: will split out.

**Comment by craigcomstock** [comment]
Ok! On a non-updated rhel-9 system I get what I would expect:

```
Installing: 'cfengine-nova-hub-3.22.0a.443209f32-24913.el9.x86_64.rpm' on 'rhel-9'
Extra Packages for Enterprise Linux 9 - x86_64  3.9 MB/s |  12 MB     00:03    
Last metadata expiration check: 0:00:03 ago on Wed 11 Jan 2023 05:26:15 PM UTC.
(try to add '--skip-broken' to skip uninstallable packages or '--nobest' to use not only best candidate packages)

Error: 
 Problem: conflicting requests
  - nothing provides selinux-policy >= 34.1.43 needed by cfengine-nova-hub-3.22.0a.443209f32-24913.el9.x86_64
```

On updating that package install succeeds.

**Comment by craigcomstock** [comment]
@cf-bottom jenkins please, full tilt!

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/9160/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/9160/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-9160/

File: packaging/cfengine-nova-hub/cfengine-nova-hub.spec.in (hunk 4 above)
Comment (vpodzime) [reply packaging/cfengine-nova-hub/cfengine-nova-hub.spec.in:15]: But it changes behavior, right? Previously `cfengine-nova-hub` would obsolete any version of `cfengine3` (whatever that is/was) and `cfengine-community`. With the above change, `cfengine-community` and `cfengine-nova-hub` can be both installed aside at the same time as long as `cfengine-community` has the same or higher version.

And the same applies to `cfengine-nova` below.

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/9165/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/9165/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-9165/

**Review by Lex-2008 — APPROVED** [review]

**Comment by craigcomstock** [comment]
All but testing-run-php-selenium-tests (known failure right now) passed, so I'm calling this good and merging it :)

**Comment by craigcomstock** [comment]
cherry picked to
3.18.x https://github.com/cfengine/buildscripts/pull/1190
3.21.x https://github.com/cfengine/buildscripts/pull/1191
